feat(react-component-annotate): Allow skipping annotations on specified components

0Calories opened the pull request:
**Note:** The large number of deleted lines is due to a lot of old tests no longer being relevant since the component ignore config system is being overhauled

Some third party libraries involve components that have compatibility issues with the `react-component-annotate` plugin, and results in errors when the `data-sentry` properties are added as props. This PR introduces a bit of a bandaid fix by introducing an option to allow users to configure the names of components that should **not** have any annotations applied.

When using the standalone Babel plugin, you can configure this option like so:

```
// babel.config.js

{
  // ... other config above ...

  plugins: [
    // Put this plugin before any other plugins you have that transform JSX code
    // The options are set by providing an object as the second element in the array, but not required
    ['@sentry/babel-plugin-component-annotate', {ignoreComponents: ['Foo', 'Bar']}]
  ],
}
```

When using the Sentry bundler plugins, you can configure this option like so:

```
sentryVitePlugin({
    // ... other options
    reactComponentAnnotation: {enabled: true, ignoreComponents: ['Foo', 'Bar']},
}),
```

TODO in future: Allow providing a source file location, since it is possible for multiple components to have the same name, but you may not want to skip annotation on all of these.

lforst commented:
@0Calories can we expedite this PR?

0Calories commented:
@lforst I'll try to finish it on friday, I believe it just needs tests

Also:
>  allow providing a source file location, since it is possible for multiple components to have the same name, but you may not want to skip annotation on all of these.

This unfortunately will have to be done in a followup PR. For now, I want to at least expose the option to skip specific components by names as that will at least provide an option for users who are struggling with incompatible components

nathan-gage commented:
will it be possible to resolve #611 with this PR -- i.e., not trying to name <Fragment/> from headless?

0Calories commented:
Hi @nathan-gage, unfortunately I don't think this would be able to help with that Headless UI issue. It's possible that you could include `React.Fragment` as a component name to be skipped, but I can't confirm whether that would work or not

0Calories commented:
@lforst This is good to go for a review now!

lforst commented on packages/bundler-plugin-core/src/types.ts:
```diff
@@ -319,6 +319,15 @@ export interface Options {
+     * A list of strings representing the names of components to ignore. The plugin will not apply `data-sentry` annotations on the DOM element for these components.
+     */
+    ignoreComponents?: string[];
+    // TODO: Support this option in the future
+    /**
+     * A list of strings representing local paths to files to ignore. The plugin will not apply `data-sentry` annotations on components in these files
+     */
+    // ignoreFiles?: string[];
```
I would just remove this comment and create an issue instead.

lforst commented on packages/dev-utils/src/generate-documentation-table.ts:
```diff
@@ -390,6 +390,21 @@ type IncludeEntry = {
+          "A list of strings representing the names of components to ignore. The plugin will not perform apply `data-sentry` annotations on the DOM element for these components.",
+        supportedBundlers: ["webpack", "vite", "rollup"],
+      },
+      // TODO: Support this option in the future
+      // {
+      //   name: "ignoreFiles",
+      //   type: "string[]",
+      //   fullDescription:
+      //     "A list of strings representing local paths to files to ignore. The plugin will not perform any annotations on components in these files.",
+      //   supportedBundlers: ["webpack", "vite", "rollup"],
+      // },
```
Same as above

lforst commented on packages/babel-plugin-component-annotate/README.md:
```diff
@@ -47,6 +47,14 @@ Using pnpm:
 
+## Options
+
+### `ignoreComponents`
```
Super nit but I think my logic is sound: I think I would probably call it `ignore*d*Components`. `ignoreComponents` sounds like imperative and that it would accept a boolean rather than a list of components.

lforst reviewed (approved):
In lgtm, I think I would call it `ignoredComponents` though. Maybe a test or two would also be dope.

0Calories commented:
Thanks for the review @lforst! Agreed on the naming of `ignoredComponents`. There's also a test I added but it's buried in the diffs because of all the obsolete tests I deleted
